core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java

```diff
@@ -0,0 +1,141 @@
+    }
+
+    @Test
+    public void testBackwardIncompatibleEntitiesAreInPreviousSpi()
+            throws Exception
+    {
+        assertThat(getPreviousSpi()).containsAll(BACKWARD_INCOMPATIBLE_CHANGES);
+    }
+
+    private static Set<String> getCurrentSpi()
+            throws IOException
+    {
+        return getSpiEntities(getSystemClassLoader(), true);
+    }
+
+    private static Set<String> getPreviousSpi()
+            throws Exception
+    {
+        try (Stream<Path> list = Files.list(Path.of("target", "released-artifacts"))) {
+            URL[] jars = list.map(TestSpiBackwardCompatibility::getUrl)
+                    .toArray(URL[]::new);
+            try (URLClassLoader urlClassLoader = new URLClassLoader(jars, getPlatformClassLoader())) {
+                return getSpiEntities(urlClassLoader, false);
+            }
+        }
+    }
+
+    private static URL getUrl(Path path)
+    {
+        try {
+            return path.toUri().toURL();
+        }
+        catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Set<String> getSpiEntities(ClassLoader classLoader, boolean includeDeprecated)
+            throws IOException
+    {
+        ImmutableSet.Builder<String> entities = ImmutableSet.builder();
+        for (ClassInfo classInfo : ClassPath.from(classLoader).getTopLevelClassesRecursive("io.trino.spi")) {
+            Class<?> clazz = classInfo.load();
+            addClassEntities(entities, clazz, includeDeprecated);
+        }
+        return entities.build();
+    }
+
+    private static void addClassEntities(ImmutableSet.Builder<String> entities, Class<?> clazz, boolean includeDeprecated)
+    {
+        if (!isPublic(clazz.getModifiers())) {
+            return;
+        }
+        entities.add("Class: " + clazz.toGenericString());
+        for (Class<?> nestedClass : clazz.getDeclaredClasses()) {
+            addClassEntities(entities, nestedClass, includeDeprecated);
+        }
+        if (!includeDeprecated && clazz.isAnnotationPresent(Deprecated.class)) {
+            return;
+        }
+        for (Constructor<?> constructor : clazz.getConstructors()) {
+            if (includeDeprecated || !constructor.isAnnotationPresent(Deprecated.class)) {
+                entities.add("Constructor: " + constructor.toGenericString());
+            }
+        }
+        for (Method method : clazz.getDeclaredMethods()) {
+            if (!isPublic(method.getModifiers())) {
+                continue;
+            }
+            if (method.getDeclaringClass().equals(Object.class)) {
+                continue;
+            }
+            if (method.getDeclaringClass().equals(Enum.class)) {
+                continue;
+            }
+            if (includeDeprecated || !method.isAnnotationPresent(Deprecated.class)) {
+                entities.add("Method: " + method.toGenericString());
+            }
```

Comment:
it would be more readable if `entities.add(` was last line in the loop body, and deprecation handling was done as all other skipping (`if (condition) continue`)